content/outcomes/secure-software-supply-chain/building-a-container.md

```diff
@@ -0,0 +1,91 @@
+
+```dockerfile
+FROM ubuntu:18.04
+COPY . /app
+RUN make /app
+ENTRYPOINT /app/my-apo
+```
+
+This Dockerfile tells the builder to use the `ubuntu:18.04` container image as a base and to copy the current directory from the local filesystem to `/app` within the container that we're building. The `RUN` command then runs the command to build the application. Finally, the `ENTRYPOINT` instruction defines what to run when the container is started. In this case, when the container is started, the executable at `/app/my-app` is ran.
+
+With that basic understanding, let's look at a more specific example. In this case, this Dockerfile is used to run a Spring Boot application:
+
+```dockerfile
+FROM openjdk:8-jdk-alpine
+EXPOSE 8080
+ARG JAR_FILE=target/my-application.jar
+ADD ${JAR_FILE} app.jar
+ENTRYPOINT ["java","-jar","/app.jar"]
+```
+
+Let's start with the two new instructions added here. Since our container will be running a web application running on port 8080, we can inform the container runtime of this by using the `EXPOSE` instruction. Additionally, from the `ADD` instruction, you might have noticed that this Dockerfile is expecting a prebuilt JAR. If you're a Java developer, you may know that different build tools, such as Maven and Gradle, build their applications differently. The `ARG` instruction defines a variable that can be used in the Dockerfile, with the option to override it when building the container. In this case, we set the default value of `JAR_FILE` to `target/my-application.jar` with the assumption that we're building this application Maven. If this is the case, the container could then just be built with the following command:
+
+```bash
+docker build -t USERNAME/my-application .
+```
+
+However, Gradle would place the build artifact in the `build/libs` directory. Luckily, the `docker build` command allows up to override this `ARG` instruction in the command line:
+
+```
+docker build --build-arg JAR_FILE=build/libs/\*.jar -t USERNAME/my-application .
+```
+
+There's many more instructions that a Dockerfile can contain. To learn more, please refer to the [Dockerfile reference](https://docs.docker.com/engine/reference/builder/) and the [best practices for writing Dockerfiles](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/).
+
+## Kaniko
+
+[Kaniko](https://github.com/GoogleContainerTools/kaniko) is an example of a tool that can help formalize your build pipeline. Where building a Dockerfile directly requires you to have Docker up and running wherever you're performing your build, Kaniko allows you to build a container from a Dockerfile on top of Kubernetes. At its core, Kaniko is a prebuilt container image with an executable that knows how to compile a Dockerfile without Docker. You still write your Dockerfile as demonstrated, but instead of running a `docker build` command, you provide it to Kaniko along with your code, and it will build and upload the container wherever you specify.
```

Comment:
The biggest advantage of Kaniko is that it does not require privilege escalation to build a container. In the context of s3c, this may be a way to ensure that you are not granting unnecessary permissions.